eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala

```diff
@@ -125,14 +126,29 @@ object NodeRelay {
       Some(InvalidOnionPayload(UInt64(2), 0))
     } else {
       payloadOut match {
-        case payloadOut: IntermediatePayload.NodeRelay.Standard =>
-          if (payloadOut.invoiceFeatures.isDefined && payloadOut.paymentSecret.isEmpty) {
-            Some(InvalidOnionPayload(UInt64(8), 0)) // payment secret field is missing
-          } else {
-            None
-          }
-        case _: IntermediatePayload.NodeRelay.ToBlindedPaths =>
-          None
+        // If we're relaying a standard payment to a non-trampoline recipient, we need the payment secret.
+        case payloadOut: IntermediatePayload.NodeRelay.Standard if payloadOut.invoiceFeatures.isDefined && payloadOut.paymentSecret.isEmpty => Some(InvalidOnionPayload(UInt64(8), 0))
+        case _: IntermediatePayload.NodeRelay.Standard => None
+        case _: IntermediatePayload.NodeRelay.ToBlindedPaths => None
+      }
+    }
+  }
+
+  /** This function identifies whether the next node is a wallet node directly connected to us, and returns its node_id. */
+  private def nextWalletNodeId(nodeParams: NodeParams, recipient: Recipient): Option[PublicKey] = {
+    recipient match {
+      // These two recipients are only used when we're the payment initiator.
+      case _: SpontaneousRecipient => None
+      case _: TrampolineRecipient => None
+      // When relaying to a trampoline node, the next node may be a wallet node directly connected to us, but we don't
+      // want to have false positives. Feature branches should check an internal DB/cache to confirm.
+      case r: ClearRecipient if r.nextTrampolineOnion_opt.nonEmpty => None
```

Comment:
This case can be extended to wake-up current Phoenix wallets: we should check our push notifications DB here and if it has an entry for the next node, we should return it, otherwise we should return None (this is a trampoline relay to a non-wallet node).